dotcom-rendering/src/lib/grid.ts

```diff
@@ -0,0 +1,164 @@
+	| 'right-column-start'
+	| 'right-column-end'
+	| 'viewport-end';
+
+const mobileColumns =
+	'[viewport-start] 0px [centre-column-start] repeat(4, 1fr) [centre-column-end] 0px [viewport-end]';
+const tabletColumns =
+	'[viewport-start] 1fr [centre-column-start] repeat(12, 40px) [centre-column-end] 1fr [viewport-end]';
+const desktopColumns =
+	'[viewport-start] 1fr [centre-column-start] repeat(8, 60px) [centre-column-end right-column-start] repeat(4, 60px) [right-column-end] 1fr [viewport-end]';
+const leftColColumns =
+	'[viewport-start] 1fr [left-column-start] repeat(2, 60px) [left-column-end centre-column-start] repeat(8, 60px) [centre-column-end right-column-start] repeat(4, 60px) [right-column-end] 1fr [viewport-end]';
+const wideColumns =
+	'[viewport-start] 1fr [left-column-start] repeat(3, 60px) [left-column-end centre-column-start] repeat(8, 60px) [centre-column-end] 60px [right-column-start] repeat(4, 60px) [right-column-end] 1fr [viewport-end]';
+const mobileColumnGap = '12px';
+const columnGap = '20px';
+
+// ----- Grid Styles ----- //
+
+const container = `
+    display: grid;
+    grid-template-columns: ${mobileColumns};
+    column-gap: ${mobileColumnGap};
+
+    ${from.mobileLandscape} {
+        column-gap: ${columnGap};
+    }
+
+    ${from.tablet} {
+        grid-template-columns: ${tabletColumns};
+    }
+
+    ${from.desktop} {
+        grid-template-columns: ${desktopColumns};
+    }
+
+    ${from.leftCol} {
+        grid-template-columns: ${leftColColumns};
+    }
+
+    ${from.wide} {
+        grid-template-columns: ${wideColumns};
+    }
+`;
+
+// ----- API ----- //
+
+/**
+ * Ask the element to span all grid columns between two grid lines. The lines
+ * can be specified either by `Line` name or by number.
+ * @param from The grid line to start from, either a `Line` name or a number.
+ * @param to The grid line to end at, either a `Line` name or a number.
+ * @returns {string} CSS to place the element on the grid.
+ *
+ * @example <caption>Will place the element in the centre column.</caption>
+ * const styles = css`
+ *   ${grid.between('centre-column-start', 'centre-column-end')}
+ * `;
+ *
+ * @example <caption>Will place the element between lines 3 and 5.</caption>
+ * const styles = css`
+ *   ${grid.between(3, 5)}
+ * `;
+ */
+const between = (from: Line | number, to: Line | number): string => `
+    grid-column: ${from} / ${to};
+`;
+
+/**
+ * Ask the element to span a number of grid columns, starting at a specific
+ * grid line. The line can be specified either by `Line` name or by number.
+ * @param start The grid line to start from, either a `Line` name or a number.
+ * @param span The number of columns to span.
+ * @returns {string} CSS to place the element on the grid.
+ *
+ * @example <caption>The element will span 3 columns from the line.</caption>
+ * const styles = css`
+ *   ${grid.span('centre-column-start', 3)}
+ * `;
+ */
+const span = (start: Line | number, span: number): string => `
```

Comment:
I appreciate the type safety that we get for named lines with this API, but I'm in two minds about the pattern. I think part of my discomfort comes from the fact that `span` is already term in the CSS spec, but we're using it here as an alias for `grid-column`. We could rename the function, but part of me thinks that we should encourage devs to use the CSS spec directly when it comes to placing elements on the grid.

There's clear value to a reusable abstraction for _creating_ the grid containers, but for placing elements on it once it's been created I'm not sure whether type safety is worth the extra abstraction for `span` and `between` etc.?

Reply:
I think that if these abstractions become helpful when you use them a lot. Maybe for this Section we don’t need them, and I’ll have a look at a pure CSS implementation.

Reply:
Merged #7146 which implements a pure CSS implementation.